sklearn/tree/_criterion.pyx

```diff
@@ -1294,111 +1278,52 @@ cdef class MAE(RegressionCriterion):
 
         Returns -1 in case of failure to allocate memory (and raise MemoryError)
         or 0 otherwise.
-        """
-        cdef intp_t i, k
-        cdef float64_t value
-        cdef float64_t weight
 
-        cdef void** left_child = self.left_child_ptr
-        cdef void** right_child = self.right_child_ptr
+        Reset might be called after an external class has changed
+        inplace self.sample_indices[start:end], hence re-computing
+        the absolute errors is needed
+        """
 
         self.weighted_n_left = 0.0
         self.weighted_n_right = self.weighted_n_node_samples
         self.pos = self.start
 
-        # reset the WeightedMedianCalculators, left should have no
-        # elements and right should have all elements.
+        for k in range(self.n_outputs - 1, -1, -1):
+            precompute_absolute_errors(
+                self.y, self.sample_weight, self.sample_indices, self.above, self.below,
+                k, self.start, self.end, self.left_abs_errors[k], self.left_medians
+            )
+            precompute_absolute_errors(
+                self.y, self.sample_weight, self.sample_indices, self.above, self.below,
+                k, self.end - 1, self.start - 1, self.right_abs_errors[k], self.right_medians
+            )
```

Comment:
Would be useful to add in-line documentation to understand what's going on here.